### ISSUE TITLE
Faster hashing

### ISSUE THREAD

**Issue by wlandau-lilly** [wlandau-lilly/drake#53]
Let's continue here with @AlexAxthelm's throughs from #4. In hashing, there is a tradeoff between the risk of collisions and speed. It is always a tricky choice, and I am hesitant to make a decision for all use cases. But what if we let the user choose the hashing algorithm? We would need more logic to accommodate multiple hashes, and it would be tricky, but it could be done. The default would stay `md5`.

**Comment by AlexAxthelm** [comment]
[Related](https://softwareengineering.stackexchange.com/a/145633)

Doesn't cover the xxhash family, but suggests that `murmur` might be a good choice.

**Comment by wlandau-lilly** [comment]
Interesting, I like that the post reports both time and collisions. I will think about it, but I am still not totally sold yet. I would have liked to see `sha1` and `md5` in that comparison. Also, murmur32 is a really short hash.

```r
digest(mtcars, algo = "murmur32")
[1] "45539bf3"
```

Not the only factor in collisions, I suspect, but it does make me skeptical.

**Comment by wlandau-lilly** [comment]
Correction: I do see discussion of `md5` and `sha*` later in the comments.

**Comment by wlandau-lilly** [comment]
I thought more about it, and I do not think we necessarily need to support multiple hashing algorithms for the same project. In the solution to #54, we could just make the choice of hash algorithm an argument to `make()`, `config()`, and `build_config()`, setting the default for the first two to "md5". There is still the matter of choosing the right default algorithm, but we can let users experiment for as long as it takes first.

**Comment by wlandau-lilly** [comment]
Thanks to `storr`, if the user accidentally changes the hashing algorithm, the error message should clear things up nicely.

```r
library(storr)
x = storr_rds("storr", hash_algorithm = "md5")
x$set("a", 1)
x = storr_rds("storr", hash_algorithm = "murmur32")
## Error: Incompatible value for hash_algorithm (existing: md5, requested: murmur32)
```

**Comment by wlandau-lilly** [comment]
On second thought, functions that interact with the `storr` cache also need the hashing algorithm specified, so it would be a pain for the user to have to set a non-md5 hashing algorithm for `cached()`, `plot_graph()`, etc. For now, it would be enough to stick to md5 (hard-coded in one place and passed to the rest of the package).

**Comment by wlandau-lilly** [comment]
## xxhash46 and sha512 for drake > 4.1.0

From r-lib/memoise#28, `memoise` uses `xxhash64` for keys that could become file names and `sha512` for everything else. @AlexAxthelm, this echoes your recommendation of `xxhash64` in #4. I am just about convinced that we should set [hash_algorithm](https://github.com/wlandau-lilly/drake/blob/master/R/hash.R#L1) to `"xxhash64"` and using `digest(..., file = TRUE, algo = "sha512")` in [rehash_file()](https://github.com/wlandau-lilly/drake/blob/master/R/hash.R#L61).

## md5 for drake <= 4.1.0

We should make `hash_algorithm` a function and use `md5` everywhere for projects previously built with `drake` <= 4.1.0. We might define another `hash_file_algorithm` function as well and have both depend on these functions:

```r
project_version <- function(){
  tryCatch(
    session(path = getwd(), search = FALSE),
    error = function(e) {
      packageVersion("drake")
    }
  )
}

use_md5 <- function(){
  version <- project_version()
  comparison <- compareVersion(version, "4.1.0")
  comparison < 1
}
```

We might have a separate `R/compatibility.R` file since more checks of this kind may be required in the future.

## Testing back-compatibility

We could do a unit test. I propose storing a 4.1.0 build of the basic example somewhere in `inst/` and copying over the files to the testing directory in the unit test. Something like this in a separate `tests/testthat/test-compatibility.R` file:

```r
context("compatibility")

test_with_dir("basic example built with 4.1.0 is still up to date", {
  # Set up a project previously built with drake 4.1.0.
  dir <- file.path("test", "built_basic_example_4.1.0")
  path <- system.file(dir, package = "drake", mustWork = TRUE)
  file.copy(path, getwd(), recursive = TRUE)
  setwd(dir)

  # Prepare to run the example with the specified testing configuration.
  opt <- test_opt()
  e <- eval(parse(text = opt$envir))
  jobs <- opt$jobs
  parallelism <- opt$parallelism
  load_basic_example(envir = e)
  my_plan <- e$my_plan
  config <- config(my_plan, envir = e,
    jobs = jobs, parallelism = parallelism,
    verbose = FALSE)

  # Check that everything is still up to date.
  testrun(config)
  expect_equal(justbuilt(), character(0))
})
```

**Comment by wlandau-lilly** [comment]
Just added a smaller back-compatibility test the `hash` branch. The cache takes up around half a MB, I downsized the example to conserve space. The package should not be bigger than 5 MB.

**Comment by wlandau-lilly** [comment]
Fixed by 50072106d6c7bea924bd67f21bbe0b0e44b25b2e. I am happy with the idea, but I also have some notes.

- I exposed the hash algorithms in `make()` in a way that stays compatible with previous projects, which opens up the possibility of performance comparison studies for hashes:

```r
make(
  ...
  short_hash_algo = ...,
  long_hash_algo = ...
)
```

- As it now says in the help file for `make()`, `short_hash_algo` is for hashes that could become the file names of internal cache/auxiliary files. `long_hash_algo`, on the other hand, does not need to obey the restrictions on the length of Windows file names, so it can be longer. This is the tradeoff from r-lib/memoise#28
- The heavy lifting for choosing hashes is done in [choose_hash_algo.R](https://github.com/wlandau-lilly/drake/blob/master/R/choose_hash_algo.R). I think I got the logic right, and I think all cases are tested because [choose_hash_algo.R](https://github.com/wlandau-lilly/drake/blob/master/R/choose_hash_algo.R) should be [fully covered in tests](https://codecov.io/github/wlandau-lilly/drake?branch=master).
- I ended up going with `sha256` rather than `sha512` because from what I read, the speed of the former is more similar to `md5` than to the latter.
- We could always switch the default to `sha512` without harming everyone's projects. Tests: [test-back-compatibility.R](https://github.com/wlandau-lilly/drake/blob/master/tests/testthat/test-back-compatibility.R) and [test-edge-cases.R](https://github.com/wlandau-lilly/drake/blob/master/tests/testthat/test-edge-cases.R). I ran these tests using [multiple.R](https://github.com/wlandau-lilly/drake/blob/master/tests/extra/multiple.R) for 4d0886a3d5d2e13ccad41f5d78e99a7a30511c5e, and I expanded the testing a little bit since.
- Similarly, users should be able to continue the projects they started in v4.1.0 without noticing any difference. ([test-back-compatibility.R](https://github.com/wlandau-lilly/drake/blob/master/tests/testthat/test-back-compatibility.R), [inst/back_compatibility/v4.1.0]([test-back-compatibility.R](https://github.com/wlandau-lilly/drake/blob/master/inst/back_compatibility/v4.1.0)))
- But this update **REALLY NEEDS TEST DRIVING**. Before CRAN, I need to do the usual rounds of major testing over and above the unit tests, plus try it out for real on a project for work.
- Otherwise, it is time for at least a patch. Because I exposed the hashes in the API, I might make the next roll-out 4.2.0 and push the next minor version updates later.

**Comment by wlandau-lilly** [comment]
@AlexAxthelm FYI: I just exposed the cache as an argument to `make()`, and I explained it in the new [storage vignette](https://github.com/wlandau-lilly/drake/blob/master/vignettes/storage.Rmd). I think this is a much better way to expose the hashes than what I had at the end of last week. Much more convenient, especially for [storr](https://github.com/richfitz/storr) users.

Regarding caches, I am also interested in adding `storr_rds()`/`RSQLite`/`redis` caches eventually, but I am currently having trouble and will probably move on to other issues for now.
